### PR TITLE
Refactor PioneerAnnotationUtils (possibly fixing build timeouts)

### PR DESCRIPTION
**Pull request by Michael1993**
Don't mind me, just trying stuff out.

### CODE REVIEW

**Comment by Michael1993** [comment]
Okay, so PioneerAnnotationUtilsTests time out probably because of deadlocking the `RepeatableFailExtension` - but I can't guarantee it because it's 1:37 AM. I'll look at this again a bit later.

**Comment by Michael1993** [comment]
Because I don't want to go through the whole log file again (I've severely underestimated how long it will be if I just log everything out):
```
2021-04-08T23:19:53.315+0000 [DEBUG] [TestEventLogger] 
2021-04-08T23:19:53.315+0000 [DEBUG] [TestEventLogger] Gradle Test Executor 1 STARTED
2021-04-08T23:19:54.494+0000 [DEBUG] [TestEventLogger] 
2021-04-08T23:19:54.494+0000 [DEBUG] [TestEventLogger] JUnitPioneer annotation utilities STARTED
2021-04-08T23:19:54.497+0000 [DEBUG] [TestEventLogger] 
2021-04-08T23:19:54.497+0000 [DEBUG] [TestEventLogger] JUnitPioneer annotation utilities > for annotations STARTED
2021-04-08T23:19:54.501+0000 [DEBUG] [TestEventLogger] 
2021-04-08T23:19:54.501+0000 [DEBUG] [TestEventLogger] JUnitPioneer annotation utilities > for annotations > that are not repeatable STARTED
2021-04-08T23:19:54.501+0000 [DEBUG] [TestEventLogger] 
2021-04-08T23:19:54.501+0000 [DEBUG] [TestEventLogger] JUnitPioneer annotation utilities > for annotations > that are not repeatable > and not stackable STARTED
2021-04-08T23:19:54.509+0000 [DEBUG] [TestEventLogger] 
2021-04-08T23:19:54.509+0000 [DEBUG] [TestEventLogger] JUnitPioneer annotation utilities > for annotations > that are repeatable STARTED
2021-04-08T23:19:54.510+0000 [DEBUG] [TestEventLogger] 
2021-04-08T23:19:54.510+0000 [DEBUG] [TestEventLogger] JUnitPioneer annotation utilities > for annotations > that are repeatable > but not stackable STARTED
2021-04-08T23:19:54.515+0000 [DEBUG] [TestEventLogger] 
2021-04-08T23:19:54.515+0000 [DEBUG] [TestEventLogger] JUnitPioneer annotation utilities > for annotations > that are not repeatable > and not stackable > discoversMethodAnnotation() STARTED
2021-04-08T23:19:54.516+0000 [DEBUG] [TestEventLogger] 
2021-04-08T23:19:54.516+0000 [DEBUG] [TestEventLogger] JUnitPioneer annotation utilities > for annotations > that are repeatable > but not stackable > discoversMethodAnnotation() STARTED
```

I think these two test methods were the ones that deadlocked with each other.

**Comment by Michael1993** [comment]
I've decided that I do not like how we test `PioneerAnnotationUtils`.

**Comment by Bukama** [comment]
Maybe a draft ;D

**Comment by Michael1993** [comment]
I need to write more tests covering the edge cases where refactoring was necessary. Also - build timeouts are not fixed yet with this.

**Comment by nipafx** [comment]
Thank you for looking into this. I'll check it out on the next stream.

**Comment by Michael1993** [comment]
Checklist for me. Cover the following refactors with tests:
* [ ] `findAnnotatedAnnotations`, new local variable `isRepeated`.
* [ ] `findOnMethod` and `findOnClass` separated
* [ ] `findOnOuterClasses` finding "too much"

**Comment by Bukama** [comment]
Note: Related to #460

**Comment by Michael1993** [comment]
I declare this ready for review. Give me some feedback on it.

File: src/main/java/org/junitpioneer/internal/PioneerAnnotationUtils.java [hunk 1]
```diff
@@ -130,11 +132,12 @@ public static boolean isAnyRepeatableAnnotationPresent(ExtensionContext context,
 	 */
 	public static <A extends Annotation> List<Annotation> findAnnotatedAnnotations(AnnotatedElement element,
 			Class<A> annotation) {
+		boolean isRepeated = annotation.isAnnotationPresent(Repeatable.class);
```
Comment (Bukama) [inline src/main/java/org/junitpioneer/internal/PioneerAnnotationUtils.java:135]: Checking if an annotation is repeatable doesn't mean that it is used multiple times, so I think the variable must be named `isRepeatable`

File: src/main/java/org/junitpioneer/internal/PioneerAnnotationUtils.java [hunk 2]
```diff
@@ -204,18 +207,52 @@ private static boolean isContainerAnnotationOf(Annotation potentialContainer, Cl
 					.orElse(Collections.emptyList());
 	}
 
+	private static <A extends Annotation> List<A> findOnClass(Class<?> element, Class<A> annotationType,
+			boolean findRepeated, boolean findAllEnclosing) {
+		if (element == null || element == Object.class)
+			return Collections.emptyList();
+		if (findRepeated)
```
Comment (Bukama) [inline src/main/java/org/junitpioneer/internal/PioneerAnnotationUtils.java:214]: this matches with this, cause you  search here for multiple usages of a repeatable annotation. If you already know that it was reapeated, then you would not need this find.

File: src/main/java/org/junitpioneer/internal/PioneerAnnotationUtils.java [hunk 3]
```diff
@@ -204,18 +207,52 @@ private static boolean isContainerAnnotationOf(Annotation potentialContainer, Cl
...
+			return AnnotationSupport.findRepeatableAnnotations(element, annotationType);
+
+		List<A> onElement = AnnotationSupport
+				.findAnnotation(element, annotationType)
+				.map(Collections::singletonList)
+				.orElse(Collections.emptyList());
+		List<A> onInterfaces = Arrays
```
Comment (Bukama) [inline src/main/java/org/junitpioneer/internal/PioneerAnnotationUtils.java:221]: Hm. You always call `onClass` and now the variable is called `onInterface`.

**Review by Bukama — APPROVED** [review]
Just some personal thoughts about variable names, but overall it looks promising.

File: src/main/java/org/junitpioneer/internal/PioneerAnnotationUtils.java (hunk 3 above)
Comment (Michael1993) [reply src/main/java/org/junitpioneer/internal/PioneerAnnotationUtils.java:221]: Will rename to `onType`

File: src/main/java/org/junitpioneer/internal/PioneerAnnotationUtils.java (hunk 1 above)
Comment (Michael1993) [reply src/main/java/org/junitpioneer/internal/PioneerAnnotationUtils.java:135]: Makes sense, will do

File: src/main/java/org/junitpioneer/internal/PioneerAnnotationUtils.java (hunk 2 above)
Comment (Michael1993) [reply src/main/java/org/junitpioneer/internal/PioneerAnnotationUtils.java:214]: Renaming this would make it inconsistent with the rest of the class, though 🤔

**Comment by nipafx** [comment]
One thing to point out that was discussed in Discord. `PioneerAnnotationUtilsTests.FindingAllAnnotations#notInheritedOnInterface` verifies that non-inherited annotations on interfaces are found when checking their implementing classes. For example:

```java
@NotInheritedAnnotation
public interface TestInterface { }

public static class Implementer implements TestInterface {

	public void notAnnotated() { }

}
```

Even though `@NotInheritedAnnotation` is not inherited, it will still be found when looking for it on `notAnnotated`. This counter-intuitive (and somewhat questionable) behavior was chosen because JUnit Jupiter does it as well and being consistent seemed important.

**Comment by nipafx** [comment]
Proposed commit message:

```
Refactor `PioneerAnnotationUtils` and tests (#457)

The hunt after build timeout causes (see #324) implicated `PioneerAnnotationUtilsTests`. Removing the Escheresque indirection and replacing the tests with a simpler approach unearthed some bugs and missing cases in `PioneerAnnotationUtils`. This change fixes them and seemingly also fixed the timeouts. We're so positive about that that we even turn the Windows builds back on.

Relates to: #324
PR: #457
```
